packages/asset-swapper/src/utils/market_operation_utils/fills.ts

```diff
@@ -71,19 +72,26 @@ function nativeOrdersToPath(
         const input = side === MarketOperation.Sell ? takerAmount : makerAmount;
         const output = side === MarketOperation.Sell ? makerAmount : takerAmount;
         const penalty = ethToOutputRate.times(fees[ERC20BridgeSource.Native] || 0);
-        const adjustedOutput = side === MarketOperation.Sell ? output.minus(penalty) : output.plus(penalty);
         const rate = makerAmount.div(takerAmount);
+        // targetInput can be less than the order size
+        // whilst the penalty is constant, it affects the adjusted output
+        // only up until the target has been exhausted.
+        // A large order and an order at the exact target should be penalized
+        // the same.
+        const clippedInput = BigNumber.min(targetInput, input);
+        // scale the clipped output inline with the input
+        const clippedOutput = clippedInput.dividedBy(input).times(output);
```

Comment:
should we be concerned about rounding errors with this operation? of does BigNumber prevent rounding issues?

Reply:
These remain very precise, but I could go either way in having these as whole `base units` which would lose precision but not necessarily have too much of an impact.